File: src/main/javascript/crypto/src/libs/UsageProofOfExponent.ts
import {ProofOfExponentInterface} from "./ProofOfExponentInterface";
import {CURVE_BN256, Point} from "./Point";
import {UsageProof} from "../asn1/shemas/ProofOfExponentASN";
import {AsnParser} from "@peculiar/asn1-schema";
import {base64ToUint8array, bnToUint8, hexStringToUint8, uint8ToBn, uint8tohex} from "./utils";
import {Asn1Der} from "./DerUtility";
import { AttestationCrypto } from "./AttestationCrypto";
import {defaultAbiCoder} from "ethers/lib/utils";

export class UsageProofOfExponent implements ProofOfExponentInterface {

    private tPoint: Point;
    private challengeResponse: bigint;
    private encoding: string;
    private nonce: Uint8Array;

    constructor() {
    }

    static fromData(tPoint: Point, challengeResponse: bigint, nonce: Uint8Array = new Uint8Array([])): UsageProofOfExponent {
        let me = new this();
        me.tPoint = tPoint;
        me.challengeResponse = challengeResponse;
        me.nonce = nonce;
        me.encoding = me.makeEncoding();
        return me;
    }

    fromBase64(base64DerEncoded: string) {
        this.encoding = base64DerEncoded;
        this.fromBytes(base64ToUint8array(base64DerEncoded));
    }

    fromBytes(bytes: Uint8Array) {
        // UsageProof

        let usageProof: UsageProof = AsnParser.parse(bytes, UsageProof);

        this.challengeResponse = uint8ToBn( new Uint8Array(usageProof.challengePoint) );
        let tPointEnc = new Uint8Array(usageProof.responseValue);
        this.nonce = new Uint8Array(usageProof.nonce);
        this.tPoint = Point.decodeFromHex(uint8tohex(tPointEnc), CURVE_BN256);
    }

    fromAbiBytes(bytes: Uint8Array){

        const parts = defaultAbiCoder.decode(["bytes", "bytes", "bytes"], bytes);

        console.log(parts);

        this.challengeResponse = uint8ToBn(hexStringToUint8(parts[0]));
        this.tPoint = Point.decodeFromHex(parts[1].substring(2), CURVE_BN256);
        this.nonce = hexStringToUint8(parts[2]);
    }

    makeEncoding() {
        let res: string = Asn1Der.encode('OCTET_STRING', uint8tohex(bnToUint8(this.challengeResponse))) +
            Asn1Der.encode('OCTET_STRING', uint8tohex(this.tPoint.getEncoded(false)))+
            Asn1Der.encode('OCTET_STRING', uint8tohex(this.nonce));
        return Asn1Der.encode('SEQUENCE_30', res);
    }


    public getPoint(): Point {
        return this.tPoint;
    }

    public getChallengeResponse(): bigint {
        return this.challengeResponse;
    }

    public getDerEncoding(): string {
        return this.encoding;
    }

    public getAbiEncoding(){
        return defaultAbiCoder.encode(
            ["bytes", "bytes", "bytes"],
            [
                bnToUint8(this.challengeResponse),
                this.tPoint.getEncoded(false),
                this.nonce
            ]
        )
    }

    public getNonce(): Uint8Array {
        return this.nonce;
    }

    public validateParameters():boolean {
        try {
            
            // Validate that point is valid on the given curve, have correct order and are not at infinity
            if (!AttestationCrypto.validatePointToCurve(this.tPoint, AttestationCrypto.curve )) {
                throw new Error("Point not in the curve");
            }

            // Check the challenge response size
            if (this.challengeResponse <= 0n || this.challengeResponse >= AttestationCrypto.curveOrder) {
                return false;
            }

            // While not strictly needed also check the point is not the generator
            if (this.tPoint.equals(AttestationCrypto.G) || this.tPoint.equals(AttestationCrypto.H)) {
                return false;
            }

            return true;
        } catch (e) {
            return false;
        }
    }
}

File: src/main/java/org/tokenscript/attestation/UsageProofOfExponent.java
package org.tokenscript.attestation;

import org.tokenscript.attestation.core.AttestationCrypto;
import org.tokenscript.attestation.core.ExceptionUtil;
import java.io.IOException;
import java.math.BigInteger;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.asn1.ASN1EncodableVector;
import org.bouncycastle.asn1.ASN1InputStream;
import org.bouncycastle.asn1.ASN1OctetString;
import org.bouncycastle.asn1.ASN1Sequence;
import org.bouncycastle.asn1.DEROctetString;
import org.bouncycastle.asn1.DERSequence;
import org.bouncycastle.math.ec.ECPoint;

public class UsageProofOfExponent implements ProofOfExponent {
  private static final Logger logger = LogManager.getLogger(UsageProofOfExponent.class);
  private final ECPoint tPoint;
  private final BigInteger challengeResponse;
  private final byte[] unpredictableNumber;
  private final byte[] encoding;

  public UsageProofOfExponent(ECPoint tPoint, BigInteger challengeResponse, byte[] unpredictableNumber) {
    this.tPoint = tPoint;
    this.challengeResponse = challengeResponse;
    this.unpredictableNumber = unpredictableNumber;
    this.encoding = makeEncoding(tPoint, challengeResponse);
  }

  public UsageProofOfExponent(ECPoint tPoint, BigInteger challengeResponse) {
    this(tPoint, challengeResponse, new byte[0]);
  }

  public UsageProofOfExponent(byte[] derEncoded) {
    this.encoding = derEncoded;
    try {
      ASN1InputStream input = new ASN1InputStream(derEncoded);
      ASN1Sequence asn1 = ASN1Sequence.getInstance(input.readObject());
      input.close();
      int asn1counter = 0;
      ASN1OctetString challengeEnc = ASN1OctetString.getInstance(asn1.getObjectAt(asn1counter++));
      this.challengeResponse = new BigInteger(challengeEnc.getOctets());
      ASN1OctetString tPointEnc = ASN1OctetString.getInstance(asn1.getObjectAt(asn1counter++));
      this.tPoint = AttestationCrypto.decodePoint(tPointEnc.getOctets());
      this.unpredictableNumber = ASN1OctetString.getInstance(asn1.getObjectAt(asn1counter++)).getOctets();
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not decode asn1", e);
    }
  }

  private byte[] makeEncoding(ECPoint tPoint, BigInteger challenge) {
    try {
      ASN1EncodableVector res = new ASN1EncodableVector();
      res.add(new DEROctetString(challenge.toByteArray()));
      res.add(new DEROctetString(tPoint.getEncoded(false)));
      res.add(new DEROctetString(unpredictableNumber));
      return new DERSequence(res).getEncoded();
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not encode asn1", e);
    }
  }

  @Override
  public ECPoint getPoint() {
    return tPoint;
  }

  @Override
  public BigInteger getChallengeResponse() {
    return challengeResponse;
  }

  @Override
  public byte[] getUnpredictableNumber() { return unpredictableNumber; }

  @Override
  public byte[] getDerEncoding() {
    return encoding;
  }

  /**
   * Verify that the values contained are correct according to RFC 8235
   * The latter part is _crucial_ in preventing attacks through edge cases.
   * NOTE: The proof itself is not verified!!!`
   * @return true if everything is ok
   */
  @Override
  public boolean verify() {
    try {
      // Validate that point is valid on the given curve, have correct order and are not at infinity
      AttestationCrypto.validatePointToCurve(tPoint, AttestationCrypto.curve);
      // Check the challenge response size
      if (challengeResponse.compareTo(BigInteger.ZERO) <= 0 || challengeResponse.compareTo(AttestationCrypto.curve.getOrder()) >= 0) {
        return false;
      }
      // While not strictly needed also check the point is not the generator
      if (tPoint.equals(AttestationCrypto.G) || tPoint.equals(AttestationCrypto.H)) {
        return false;
      }
      return true;
    } catch (SecurityException e) {
      return false;
    }
  }
}
